hail/src/main/scala/is/hail/expr/ir/lowering/LowerBlockMatrixIR.scala

```diff
@@ -184,30 +186,695 @@ abstract class BlockMatrixStage(val letBindings: IndexedSeq[(String, IR)], val b
+
+    if (!trivialCtxs) {
+      val (binding, contexts) = _contexts.memo
+      (_letBindings :+ binding, contexts)
+    } else {
+      (_letBindings, _contexts)
+    }
+  }
+
+  def print(ctx: ExecuteContext): Unit = {
+    println(s"contexts:\n${contexts.print(ctx)}\nbody(${ctxRefName}) = ${Pretty(ctx, _blockIR)}")
+  }
+
+  def blockIR(ctx: Ref): IR = {
+    if (ctx.name == ctxRefName)
+      _blockIR
+    else
+      Let(ctxRefName, ctx, _blockIR)
+  }
+
+  def ctxType: Type = contexts.elementType
+
+  def ctxRef: IR = Ref(ctxRefName, ctxType)
+
+  def toOldBMS: BlockMatrixStage = {
+    new BlockMatrixStage(letBindings, broadcastVals.toArray, ctxType) {
+      override def blockContext(idx: (Int, Int)): IR = contexts(idx._1, idx._2)
+
+      override def blockBody(ctxRef: Ref): IR =
+        Let(ctxRefName, ctxRef, _blockIR)
+    }
+  }
 
-        val elt = bindIR(lowered.blockContext(rowBlock -> colBlock)) { ctx =>
-          NDArrayRef(lowered.blockBody(ctx), FastIndexedSeq(I64(iInBlock), I64(jInBlock)), -1)
+  def getBlock(i: IR, j: IR): IR = Let(ctxRefName, contexts(i, j), _blockIR)
+
+  def getElement(i: IR, j: IR): IR = {
+    assert(i.typ == TInt64)
+    assert(j.typ == TInt64)
+    val blockSize = typ.blockSize.toLong
+    bindIR(i floorDiv blockSize) { rowBlock =>
+      bindIR(j floorDiv blockSize) { colBlock =>
+        val iInBlock = i - rowBlock * blockSize
+        val jInBlock = j - colBlock * blockSize
+
+        val elt = NDArrayRef(getBlock(rowBlock.toI, colBlock.toI), FastIndexedSeq(iInBlock, jInBlock), -1)
+
+        wrapLetsAndBroadcasts(elt)
+      }
+    }
+  }
+
+  def transposed: BlockMatrixStage2 = {
+    val newBlockIR = NDArrayReindex(_blockIR, FastIndexedSeq(1, 0))
+    new BlockMatrixStage2(letBindings, broadcastVals, typ.transpose, BMSContexts.transpose(contexts, typ), ctxRefName, newBlockIR)
+  }
+
+  private def wrapLetsAndBroadcasts(ctxIR: IR): IR = {
+    (letBindings ++ broadcastVals).foldRight[IR](ctxIR) { case ((f, v), accum) => Let(f, v, accum) }
+  }
+
+  def densify: BlockMatrixStage2 = contexts match {
+    case _: DenseContexts => this
+    case contexts: SparseContexts =>
+      val newContexts = contexts.mapDense { (i, j, oldContext) =>
+        val (m, n) = typ.blockShapeIR(i, j)
+        makestruct("oldContext" -> oldContext, "nRows" -> m, "nCols" -> n)
+      }
+      def newBlock(context: Ref): IR = {
+        bindIR(GetField(context, "oldContext")) { oldContext =>
+          If(IsNA(oldContext),
+            MakeNDArray.fill(
+              zero(typ.elementType),
+              FastIndexedSeq(GetField(oldContext, "nRows"), GetField(oldContext, "nCols")),
+              False()),
+            blockIR(oldContext))
         }
+      }
+      BlockMatrixStage2(letBindings, broadcastVals, typ, newContexts, newBlock)
+  }
+
+  def withSparsity(rowPos: IR, rowIdx: IR, newType: BlockMatrixType, isSubset: Boolean = false): BlockMatrixStage2 = {
+    val (newLetBindings, rowPosRef, rowIdxRef) = {
```

Comment:
can we add a bailout `return this` if the newType is the same as the old type? In practice we generate a lot of map2s where the old sparsity is the new, I think, and this can keep IR size down.